File: partiql-plan/src/main/java/org/partiql/plan/rex/RexPathSymbol.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;

import java.util.List;

/**
 * Logical path by symbol lookup expression abstract base class.
 */
public abstract class RexPathSymbol extends RexBase {

    /**
     * @return new RexPathSymbol instance
     */
    @NotNull
    public static RexPathSymbol create(@NotNull Rex operand, @NotNull String symbol) {
        return new Impl(operand, symbol);
    }

    /**
     * @return operand rex (operand 0)
     */
    @NotNull
    public abstract Rex getOperand();

    /**
     * @return symbol string
     */
    @NotNull
    public abstract String getSymbol();

    @Override
    @NotNull
    protected final RexType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getOperand());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitPathSymbol(this, ctx);
    }

    private static class Impl extends RexPathSymbol {

        private final Rex operand;
        private final String symbol;

        private Impl(@NotNull Rex operand, @NotNull String symbol) {
            this.operand = operand;
            this.symbol = symbol;
        }

        @NotNull
        @Override
        public Rex getOperand() {
            return operand;
        }

        @NotNull
        @Override
        public String getSymbol() {
            return symbol;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexPathSymbol.kt
package org.partiql.plan.rex

/**
 * Logical operator for path lookup by symbol.
 */
public interface RexPathSymbol : Rex {

    public fun getOperand(): Rex

    public fun getSymbol(): String

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitPathSymbol(this, ctx)
}

/**
 * Standard internal implementation for [RexPathSymbol].
 */
internal class RexPathSymbolImpl(operand: Rex, symbol: String, type: RexType) : RexPathSymbol {

    // DO NOT USE FINAL
    private var _operand = operand
    private var _symbol = symbol
    private var _type = type

    override fun getOperand() = _operand

    override fun getSymbol() = _symbol

    override fun getType(): RexType = _type

    override fun getChildren(): Collection<Rex> = listOf(_operand)
}
